Add URL list drag and drop example

Susurrus opened the pull request:
This example appears to work correctly for me now. I think this was resolved by fixing internal list structures are some point, so I wanted to get this queued up for merging. This requires a new dependency, but hopefully that's not an issue.

This is something that applications likely want to do (like one I'm working on) which is why I created this example in the first place. So I think it'd be helpful to other devs in addition to the basic drag-and-drop example.

GuillaumeGomez commented on src/bin/drag_and_drop_textview.rs:
```diff
@@ -0,0 +1,61 @@
+
+    // GTK & main window boilerplate
+    window.connect_delete_event(|_, _| {
+                                    gtk::main_quit();
```
Urg. Ident a bit ugly in here.

Susurrus replied:
That was done by rustfmt. Would you prefer I indent regularly?

GuillaumeGomez replied:
Yes please!

Susurrus commented:
Sure, I'll expand with some comments and more whitespace.

EPashkin commented:
It still have problem, described in https://github.com/gtk-rs/examples/pull/119#issuecomment-298149521 in Windows. Don't tested on Ubuntu

Susurrus commented:
@GuillaumeGomez Could you try out this example and see if you have the problem as well?

GuillaumeGomez commented:
I'll when I have time, so hopefully this evening or tomorrow. Don't hesitate to ping me if I forget.

Susurrus commented:
@GuillaumeGomez I also addressed your whitespace and code comments request.

GuillaumeGomez commented:
:+1: 

Thanks!

EPashkin commented:
On Ubuntu it also `drag_data_received` comes 2 times per drop.
Also it failing with many `Gtk-Critical` warnings on dropping 2 files
or with `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RelativeUrlWithoutBase', /checkout/src/libcore/result.rs:859`

Hack in https://github.com/gtk-rs/examples/pull/119#issuecomment-298156793 fixes on both Ubuntu and Windows

EPashkin commented:
Ping, @Susurrus

Susurrus commented:
@EPashkin I guess I've been waiting to see what you want me to do here. Implementing a hack in the application doesn't seem like the right path forward. I wrote a C version of this I think in #119, did it have the same problem? If so, I should file an upstream bug report, yes?

EPashkin commented:
On Windows:
`drag_and_drop.c` print only single pair of `drag-data-get/drag-data-received` per drop attempt
and current drag_and_drop.rs do same,
cross dropping won't works for both version for me.
drag_and_drop_textview.rs still have [bug](https://github.com/gtk-rs/examples/pull/119#issuecomment-298149521).

Filling upstream bug report is good, maybe we just pass wrong parameters to code.

Susurrus commented:
Alright, I've rebased on the latest `pending` branch and also updated to use the `Application` framework. This is working for me nicely on Linux 4.13 running GTK 3.22 on Fedora 27. @EPashkin Are you still seeing weird behavior on Ubuntu or Windows?

EPashkin commented:
@Susurrus Yes, Bug still here, at minimum on Windows10, and my hack still fix it.
I will test on Ubuntu today.

Susurrus commented:
@EPashkin Did you ever get to test it on Ubuntu?

I also created a C version with almost identical functionality and it works fine for me. Can you test it for you as well?
```C
//gcc -Wall drag_and_drop.c -o drag_and_drop `pkg-config --cflags --libs gtk+-3.0`
#include <stdio.h>
#include <string.h>
#include <gtk/gtk.h>
#include <gdk/gdk.h>

static void drag_data_received(GtkWidget *widget, GdkDragContext *d, gint32 i, gint32 j, GtkSelectionData *s, guint32 x, guint32 y)
{
  GtkTextBuffer *buffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(widget));
  gtk_text_buffer_set_text(buffer, "", -1);

  gchar **uris = gtk_selection_data_get_uris(s);
  for (gchar **uris_iter = uris; uris_iter && *uris_iter; ++uris_iter) {
    gtk_text_buffer_insert_at_cursor(buffer, *uris_iter, -1);
  }
  g_strfreev(uris);
}

static void
activate (GApplication *app,
          gpointer      user_data)
{
  GtkWidget *window = gtk_application_window_new (GTK_APPLICATION (app));
  gtk_window_set_title(GTK_WINDOW(window), "Drag and Drop Example with a TextView");

  GtkWidget *label = gtk_label_new("Drag files and/or folders onto the TextView below.");

  GtkWidget *text_view = gtk_text_view_new();
  gtk_text_view_set_wrap_mode(GTK_TEXT_VIEW(text_view), GTK_WRAP_WORD);
  gtk_text_view_set_cursor_visible(GTK_TEXT_VIEW(text_view), FALSE);
  g_signal_connect(text_view, "drag-data-received", G_CALLBACK(drag_data_received), NULL);

  GtkWidget *scrolled_text_view = gtk_scrolled_window_new(NULL, NULL);
  gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(scrolled_text_view), GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);
  gtk_container_add(GTK_CONTAINER(scrolled_text_view), text_view);

  GtkTargetEntry targets[] = {
      {"text/uri-list", GTK_TARGET_OTHER_APP, 0},
  };
  gtk_drag_dest_set(text_view, GTK_DEST_DEFAULT_ALL, targets, 1, GDK_ACTION_COPY);

  GtkWidget *vbox = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
  gtk_box_pack_start(GTK_BOX(vbox), label, FALSE, FALSE, 0);
  gtk_box_pack_start(GTK_BOX(vbox), scrolled_text_view, TRUE, TRUE, 0);

  gtk_container_add(GTK_CONTAINER(window), vbox);
  gtk_widget_show_all(window);
}

int
main (int argc, char **argv)
{
  GtkApplication *app;
  int status;

  app = gtk_application_new ("org.gtk-rs.examples.drag_and_drop_textview", G_APPLICATION_FLAGS_NONE);
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  status = g_application_run (G_APPLICATION (app), argc, argv);
  g_object_unref (app);

  return status;
}
```

EPashkin commented:
@Susurrus can you test c version with log on Fedora?

Susurrus commented:
I see the correct behavior on Fedora, but I don't know how to get the debugging info printed to the terminal. I tried running `G_DEBUG=all ./drag_and_drop_textview`, which I thought would do it. What's the correct incantation?

EPashkin commented:
`set G_MESSAGES_DEBUG=all ./drag_and_drop_textview`

EPashkin commented:
without set

Susurrus commented:
So I ran it and when I dragged 2 folders from Nautilus into the window I saw the expected output:
```
$ G_MESSAGES_DEBUG=all ./drag_and_drop_textview 
(drag_and_drop_textview:31626): Gtk-DEBUG: Connecting to session manager
** (drag_and_drop_textview:31626): DEBUG: received 1
** (drag_and_drop_textview:31626): DEBUG: file:///home/susurrus/Projects
** (drag_and_drop_textview:31626): DEBUG: file:///home/susurrus/Public
```

Susurrus commented:
And then when I dragged a single text file onto it I also saw expected behavior (though I don't know why the name is so manged when printed, should have printed "Google tpm presentation.txt" at the end, but the text in the window is correct):
```
** (drag_and_drop_textview:31626): DEBUG: received 2
** (drag_and_drop_textview:31626): DEBUG: file:///home/susurrus/Google               (nil)m            0x984dc8resentation.txt
```

EPashkin commented:
Bad, then bug only Ubuntu related :(
I get `DEBUG: received 1 .. DEBUG: received 2..` on each drop.

Susurrus commented:
But it happens on Windows too, which seems odd. Should a bug be filed against Ubuntu or GTK+ as well?

Susurrus commented:
I'd be happy to file a bug with GTK+, but I don't want to file an Ubuntu bug, since I don't have any Ubuntu systems. I know @GuillaumeGomez was interested in testing this out sometime, but I don't know what platform he's on. Maybe he can reproduce the bug you're seeing as well.

GuillaumeGomez commented:
I don't have an ubuntu either. I'm on mint, which is based on debian, just like ubuntu. Maybe the bug is the same there?

EPashkin commented:
Theoretically yes, but as not you not me on Ubuntu it may tough

Susurrus commented:
Maybe, worth testing out if/when you have time, both the C version and the Rust version (their behavior *should* match). Oh, are any of you running Wayland or X11? I'm on Wayland with Gnome Shell and Nautilus.

I think I'll hop on IRC with the Gnome guys today and see if anyone can reproduce the bug we're seeing with the C code. Because the Rust implementation here seems correct, it just boils down to how many platforms are affected by this GTK+ bug, if that's indeed what it is. Seems weird it doesn't occur on Fedora.

GuillaumeGomez commented:
X11 I suppose.

Susurrus commented:
@EPashkin I found the exact problem we're running into posted on the GTK+ bug tracker: https://bugzilla.gnome.org/show_bug.cgi?id=150141 The workaround is in these examples to specify `0` instead of `DEST_DEFAULT_ALL` in the call to `drag_dest_set`. Can you try that in the C and Rust version and see if it works in both?

Susurrus commented:
If so, we'll put in the workaround and link to the bug once I either reopen that old one or file a new one.

EPashkin commented:
@Susurrus Using `text_view.drag_dest_set(DestDefaults::empty(), &targets, DragAction::COPY);` works on Windows 10 and Ubuntu.

Susurrus commented:
I've updated [GTK+ bug 150141](https://bugzilla.gnome.org/show_bug.cgi?id=150141) with the latest info.

Susurrus commented:
It's actually not clear to me if I should have even been using `DestDefaults::ALL` at the start here. I changed it to `HIGHLIGHT`, which doesn't do anything for me on Fedora, but hopefully still gives some good UI feedback on other platforms. @EPashkin Let me know if you think this is good.

EPashkin commented:
@Susurrus Works fine for all 4 cases.

Susurrus commented:
Anything else I need for this example to make it mergeable?

EPashkin commented:
@Susurrus All good for me. Thanks.

@GuillaumeGomez 👍

GuillaumeGomez commented:
Thanks a lot @Susurrus for your work in here and thanks a lot @EPashkin for helping @Susurrus!
